### PR TITLE
CI autogenerate and commit parser.c

### LINKED ISSUE #76: generated files and workflows
We [need to check in the generated parser files](https://github.com/DerekStride/tree-sitter-sql/pull/59) to support downstream use, but that gets us into merge conflicts every time someone cuts a new branch before someone else lands theirs. It's only happened a few times so far, but if the pace picks up we should consider an action that can build those on pull requests and squash merge.

We could commit after build+test on the branch in progress, but that seems like a bit much.

### PR DESCRIPTION
**Pull request by matthias-Q**
This is a suggestion for resolve #76 

I tried to test it on my branch and I think it works. We also need to add the `parser.c` in the `.gitignore`

### CODE REVIEW

**Comment by matthias-Q** [comment]
Here are some references:
https://github.com/orgs/community/discussions/26615
https://stackoverflow.com/questions/58139406/only-run-job-on-specific-branch-with-github-actions
https://github.blog/changelog/2021-02-08-github-actions-skip-pull-request-and-push-workflows-with-skip-ci/

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -21,3 +21,19 @@ jobs:
...
       - run: npm test
+  deploy:
+    runs-on: ubuntu-latest
+    needs: test
+    if: github.ref == 'refs/heads/main'
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-node@v2
+        with:
+          node-version: 14
```
Comment (LeoniePhiline) [inline .github/workflows/ci.yml:32]: Security support ends in 2 months and 3 weeks
(30 Apr 2023)

https://endoflife.date/nodejs

IIRC node 18 (latest LTS) worked for you in building node-gyp? That should be used (Support ends in 2 years and 2 months
(30 Apr 2025))

Comment (matthias-Q) [reply .github/workflows/ci.yml:32]: So you suggest this?
```yaml
- uses: actions/checkout@v3
- uses: actions/setup-node@v3
  with:
    node-version: 18
```

I saw the deprecation warning, but it stated Node12 not Node14

Comment (LeoniePhiline) [reply .github/workflows/ci.yml:32]: Yes. And I suggest changing the node version a few lines above as well.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -21,3 +21,19 @@ jobs:
...
+      - run: npm install
+      - run: |
+          git config --global user.name "github-actions[bot]"
+          git config --global user.email "41898282+github-actions[bot]@users.noreply.github.com"
+          git add src/parser.c
```
Comment (matthias-Q) [inline .github/workflows/ci.yml:37]: We should probably also include
```yaml
    git add src/grammar.json
    git add src/node-types.json
```

Comment (LeoniePhiline) [reply .github/workflows/ci.yml:37]: Just `git add src`

Comment (LeoniePhiline) [inline .github/workflows/ci.yml:37]:
```suggestion
          git add src
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -21,3 +21,19 @@ jobs:
...
+          git commit -m "[skip ci] auto generate parser.c"
```
Comment (LeoniePhiline) [inline .github/workflows/ci.yml:38]:
```suggestion
          git commit -m "[skip ci] auto generate parser"
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -21,3 +21,19 @@ jobs:
           node-version: 14
```
Comment (LeoniePhiline) [inline .github/workflows/ci.yml]:
```suggestion
          node-version: 18
```

**Comment by dmfay** [comment]
Unfortunately I think there is an issue with regenerating the parser in a new commit on `main`: there will be a (brief) window during which the state of `main` is inconsistent and consumers could wind up with unexpected behavior.

I think the ideal place to regenerate the files is in a final commit on the branch _before_ it's merged into `main`; I've done [a little experimenting with a label-based process on my fork](https://github.com/dmfay/tree-sitter-sql/blob/main/.github/workflows/merge.yml) but haven't got it working yet -- feel free to pick it up and run with it if you think it's promising!

**Comment by matthias-Q** [comment]
Ah, I see. You are right, for the 2-3 minutes while the CI actions are running, we have an inconsistent code base.

Can you explain how your approach should work in theory? If a PR is accepted, it should be labled with `merge` and then it triggers the actions in `merge.yaml`, correct?

**Comment by dmfay** [comment]
> Ah, I see. You are right, for the 2-3 minutes while the CI actions are running, we have an inconsistent code base.
> 
> Can you explain how your approach should work in theory? If a PR is accepted, it should be labeled with `merge` and then it triggers the actions in `merge.yaml`, correct?

yes, that's the idea! Once labeled it should merge `main` into the branch (without committing), check out the `main` version of the parser files (eliminating any conflicts outside grammar.js which must be manually resolved), rebuild them to include the branch changes, then commit, push, and merge the pull request, and finally clean up the label.

**Comment by matthias-Q** [comment]
I think that your solution is better. What is currently not working?

**Comment by matthias-Q** [comment]
I was looking in you version and into https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#using-conditionals

Did you tried this line with out the curly braces `{{...}}`? 
```
    if: github.event.label.name == 'merge'
```

**Comment by matthias-Q** [comment]
@dmfay @DerekStride I think I solved the issue. Building upon @dmfay work, I added a `merge.yml`, which triggers a "push force" when the `merge` label has been let to the PR.

Unfortunately, I uses a third party Github Action: https://github.com/marketplace/actions/github-push

I have tested this on a private repository. You also need to allow the Github actions bot to have write permissions on the repo.

**Comment by DerekStride** [comment]
I took a different approach in #100, I like the label approach if we still want the artifacts on the `main` branch but I think I'd prefer to keep them separate.

**Comment by DerekStride** [comment]
I'm going to close this since #100 was merged, we can re-open it if we decide that #100 was the wrong decision in the future.

**Pull request closed without merging**